src/rust/session/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+
+    /// Get the set of readiness events the session is waiting for
+    fn readiness(&self) -> Interest {
+        if self.write_buffer.is_empty() {
+            Interest::READABLE
+        } else {
+            Interest::READABLE | Interest::WRITABLE
+        }
+    }
+
+    /// Returns a boolean which indicates if the session is handshaking
+    pub fn is_handshaking(&self) -> bool {
+        self.stream.is_handshaking()
+    }
+
+    /// Drives the handshake for the session. A successful result indicates that
+    /// the session hadshake is completed successfully. The error result should
+    /// be checked to determine if the operation would block, resulted in some
+    /// unrecoverable error, or if the session was not in a handshaking state
+    /// when this was called.
+    pub fn do_handshake(&mut self) -> Result<(), std::io::Error> {
+        self.stream.do_handshake()
+    }
+
+    /// Closes the session and the underlying stream.
+    pub fn close(&mut self) {
+        self.stream.close();
+    }
+
+    /// Returns the number of bytes in the read buffer
+    pub fn read_pending(&self) -> usize {
+        self.read_buffer.len()
+    }
+
+    /// Returns the number of bytes in the write buffer
+    pub fn write_pending(&self) -> usize {
+        self.write_buffer.len()
+    }
+
+    pub fn write_capacity(&self) -> usize {
+        self.capacity.saturating_sub(self.write_pending())
+    }
+
+    /// Returns a reference to the internally buffered data.
+    ///
+    /// Unlike [`fill_buf`], this will not attempt to fill the buffer if it is
+    /// empty.
+    ///
+    /// [`fill_buf`]: BufRead::fill_buf
+    pub fn buffer(&self) -> &[u8] {
+        self.read_buffer.borrow()
+    }
+
+    pub fn peer_addr(&self) -> Result<SocketAddr, std::io::Error> {
+        self.stream.peer_addr()
+    }
+}
+
+impl Read for Session {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error> {
+        if self.read_buffer.is_empty() {
+            self.fill_buf()?;
+        }
+        let bytes = std::cmp::min(buf.len(), self.read_buffer.len());
+        let buffer: &[u8] = self.read_buffer.borrow();
+        buf[0..bytes].copy_from_slice(&buffer[0..bytes]);
+        self.consume(bytes);
+        Ok(bytes)
+    }
+}
+
+impl BufRead for Session {
+    fn fill_buf(&mut self) -> Result<&[u8], std::io::Error> {
+        increment_counter!(&Stat::SessionRecv);
+        let mut total_bytes = 0;
+        loop {
+            self.read_buffer.reserve(self.capacity);
+            match self.stream.read(self.read_buffer.borrow_mut()) {
+                Ok(0) => {
+                    // Stream is disconnected, stop reading
+                    break;
+                }
+                Ok(bytes) => {
+                    // self.read_buffer.extend(&self.tmp_buffer[0..bytes]);
+                    total_bytes += bytes;
+                    if bytes < self.capacity {
+                        // we read less than the temp buffer size, next read
```

Comment:
in this comment line: `temp` -> `read`

Reply:
Not strictly true? Here `self.capacity` is serving as a read chunk size - this might not be a good strategy here anyway - it might be worth an extra call to read() and ensure we get a `WouldBlock` rather than assume that read() resulting in fewer bytes than our chunk size is a hint that we'll get a `WouldBlock` on the next read().

Thoughts?

Reply:
removed this conditional in favor of an extra read to ensure we get a `WouldBlock`